compiler/rustc_data_structures/src/sync.rs

```diff
@@ -26,6 +26,36 @@ use std::panic::{catch_unwind, resume_unwind, AssertUnwindSafe};
 pub use std::sync::atomic::Ordering;
 pub use std::sync::atomic::Ordering::SeqCst;
 
+mod mode {
+    use super::Ordering;
+    use std::sync::atomic::AtomicU8;
+
+    const UNINITIALIZED: u8 = 0;
+    const SINGLE_THREAD: u8 = 1;
+    const MULTIPLE_THREADS: u8 = 2;
+
+    static PARALLEL: AtomicU8 = AtomicU8::new(UNINITIALIZED);
+
+    #[inline]
+    pub fn active() -> bool {
+        match PARALLEL.load(Ordering::Relaxed) {
+            SINGLE_THREAD => false,
+            MULTIPLE_THREADS => true,
+            _ => panic!("uninitialized parallel mode!"),
+        }
+    }
+
+    // Only set by the `-Z threads` compile option
+    pub fn set_parallel_mode(parallel: bool) {
+        let set: u8 = if parallel { MULTIPLE_THREADS } else { SINGLE_THREAD };
+        let previous =
+            PARALLEL.compare_exchange(UNINITIALIZED, set, Ordering::Relaxed, Ordering::Relaxed);
+        assert!(previous.is_ok() || previous == Err(set));
+    }
+}
+
+pub use mode::{active, set_parallel_mode};
+
 cfg_if! {
     if #[cfg(not(parallel_compiler))] {
         pub auto trait Send {}
```

Comment:
These dummy traits could be removed now.

Reply:
Done